### ISSUE TITLE
kopf progress not being set on annotations or status

### ISSUE THREAD

**Issue by chamcca** [nolar/kopf#819]
### Long story short

Using a very simple `on.create` event handler, the progress is not being stored on the resources' annotation or status. This results in every resource firing an `on.resume` when the operator is restarted. Debug logs show the resource being patched, but there are no progress annotations or status attributes in the patch.

In addition, if the operator has two `on.create` event handlers, debug logs show both progress annotations and status values being patched when the first event handler completes, but these values are then removed when the second event handler fires. The order of the handlers does not matter, the second one to fire always removes the progress values. Attempting to use a custom AnnotationProgressPrefix or StatusProgressStorage results in the same behavior.

Logs and code included.

### Kopf version

1.33.0

### Kubernetes version

1.20

### Python version

3.7

### Code

```python
# Single on.create event handler example

import kopf

@kopf.on.create("imagereplications")
def created1(spec: kopf.Spec, logger: kopf.Logger, **_: Any) -> None:
    logger.info("Created1")
    return

#####################################

# Example with 2 on.create event handlers

import kopf

@kopf.on.create("imagereplications")
def created1(spec: kopf.Spec, logger: kopf.Logger, **_: Any) -> None:
    logger.info("Created1")
    return

@kopf.on.create("imagereplications")
def created2(spec: kopf.Spec, logger: kopf.Logger, **_: Any) -> None:
    logger.info("Created2")
    return
```

### Logs

```none
# Logs resulting from operator with a single on.create event handler 
# Notice the single "Patching with..." does not include progress annotations or status

[2021-08-18 20:14:00,319] kopf._core.reactor.r [DEBUG   ] Starting Kopf 1.33.0.
[2021-08-18 20:14:00,319] kopf._core.engines.a [INFO    ] Initial authentication has been initiated.
[2021-08-18 20:14:00,319] kopf.activities.auth [DEBUG   ] Activity 'login_via_client' is invoked.
[2021-08-18 20:14:00,320] kopf._core.engines.p [DEBUG   ] Serving health status at http://0.0.0.0:8080/healthz
[2021-08-18 20:14:00,321] kopf.activities.auth [DEBUG   ] Client is configured in cluster with service account.
[2021-08-18 20:14:00,321] kopf.activities.auth [INFO    ] Activity 'login_via_client' succeeded.
[2021-08-18 20:14:00,321] kopf._core.engines.a [INFO    ] Initial authentication has finished.
[2021-08-18 20:14:00,351] kopf._cogs.clients.w [DEBUG   ] Starting the watch-stream for customresourcedefinitions.v1.apiextensions.k8s.io cluster-wide.
[2021-08-18 20:14:00,351] kopf._cogs.clients.w [DEBUG   ] Starting the watch-stream for imagereplications.v1.orbit.aws cluster-wide.
[2021-08-18 20:14:08,767] kopf.objects         [DEBUG   ] [orbit-system/image-replication-z6sfr] Creation is in progress: {'apiVersion': 'orbit.aws/v1', 'kind': 'ImageReplication', 'metadata': {'creationTimestamp': '2021-08-18T20:14:08Z', 'generateName': 'image-replication-', 'generation': 1, 'managedFields': [{'apiVersion': 'orbit.aws/v1', 'fieldsType': 'FieldsV1', 'fieldsV1': {'f:metadata': {'f:generateName': {}}, 'f:spec': {'.': {}, 'f:destination': {}, 'f:source': {}}}, 'manager': 'kubectl-create', 'operation': 'Update', 'time': '2021-08-18T20:14:08Z'}], 'name': 'image-replication-z6sfr', 'namespace': 'orbit-system', 'resourceVersion': '5847689', 'uid': '136c392b-4f4e-4999-9aa3-c0ba58c94699'}, 'spec': {'destination': '011316101677.dkr.ecr.us-west-2.amazonaws.com/orbit-dev-env/nginx:latest', 'source': 'nginx:latest'}}
[2021-08-18 20:14:08,767] kopf.objects         [DEBUG   ] [orbit-system/image-replication-z6sfr] Handler 'created1' is invoked.
[2021-08-18 20:14:08,768] kopf.objects         [INFO    ] [orbit-system/image-replication-z6sfr] Created
[2021-08-18 20:14:08,770] kopf.objects         [INFO    ] [orbit-system/image-replication-z6sfr] Handler 'created1' succeeded.
[2021-08-18 20:14:08,770] kopf.objects         [INFO    ] [orbit-system/image-replication-z6sfr] Creation is processed: 1 succeeded; 0 failed.
[2021-08-18 20:14:08,771] kopf.objects         [DEBUG   ] [orbit-system/image-replication-z6sfr] Patching with: {'metadata': {'annotations': {'kopf.zalando.org/last-handled-configuration': '{"spec":{"destination":"011316101677.dkr.ecr.us-west-2.amazonaws.com/orbit-dev-env/nginx:latest","source":"nginx:latest"}}\n'}}}
[2021-08-18 20:14:08,885] kopf.objects         [DEBUG   ] [orbit-system/image-replication-z6sfr] Something has changed, but we are not interested (the essence is the same).
[2021-08-18 20:14:08,885] kopf.objects         [DEBUG   ] [orbit-system/image-replication-z6sfr] Handling cycle is finished, waiting for new changes.


#####################################################

# Logs resulting from operator with 2 on.create event handlers 
# Notice the two "Patching with..." entries, the first includes annotation and status progress values
# The second "Patching with.. " sets the annotation and status progress values to None

[2021-08-18 20:15:34,318] kopf._core.reactor.r [DEBUG   ] Starting Kopf 1.33.0.
[2021-08-18 20:15:34,318] kopf._core.engines.a [INFO    ] Initial authentication has been initiated.
[2021-08-18 20:15:34,318] kopf.activities.auth [DEBUG   ] Activity 'login_via_client' is invoked.
[2021-08-18 20:15:34,320] kopf._core.engines.p [DEBUG   ] Serving health status at http://0.0.0.0:8080/healthz
[2021-08-18 20:15:34,320] kopf.activities.auth [DEBUG   ] Client is configured in cluster with service account.
[2021-08-18 20:15:34,320] kopf.activities.auth [INFO    ] Activity 'login_via_client' succeeded.
[2021-08-18 20:15:34,320] kopf._core.engines.a [INFO    ] Initial authentication has finished.
[2021-08-18 20:15:34,343] kopf._cogs.clients.w [DEBUG   ] Starting the watch-stream for customresourcedefinitions.v1.apiextensions.k8s.io cluster-wide.
[2021-08-18 20:15:34,344] kopf._cogs.clients.w [DEBUG   ] Starting the watch-stream for imagereplications.v1.orbit.aws cluster-wide.
[2021-08-18 20:15:46,545] kopf.objects         [DEBUG   ] [orbit-system/image-replication-929js] Creation is in progress: {'apiVersion': 'orbit.aws/v1', 'kind': 'ImageReplication', 'metadata': {'creationTimestamp': '2021-08-18T20:15:46Z', 'generateName': 'image-replication-', 'generation': 1, 'managedFields': [{'apiVersion': 'orbit.aws/v1', 'fieldsType': 'FieldsV1', 'fieldsV1': {'f:metadata': {'f:generateName': {}}, 'f:spec': {'.': {}, 'f:destination': {}, 'f:source': {}}}, 'manager': 'kubectl-create', 'operation': 'Update', 'time': '2021-08-18T20:15:46Z'}], 'name': 'image-replication-929js', 'namespace': 'orbit-system', 'resourceVersion': '5848830', 'uid': 'a7887a0f-b599-4ab6-8760-e1c8d71a228c'}, 'spec': {'destination': '011316101677.dkr.ecr.us-west-2.amazonaws.com/orbit-dev-env/nginx:latest', 'source': 'nginx:latest'}}
[2021-08-18 20:15:46,545] kopf.objects         [DEBUG   ] [orbit-system/image-replication-929js] Handler 'created1' is invoked.
[2021-08-18 20:15:46,546] kopf.objects         [INFO    ] [orbit-system/image-replication-929js] Created1
[2021-08-18 20:15:46,548] kopf.objects         [INFO    ] [orbit-system/image-replication-929js] Handler 'created1' succeeded.
[2021-08-18 20:15:46,548] kopf.objects         [DEBUG   ] [orbit-system/image-replication-929js] Patching with: {'metadata': {'annotations': {'kopf.zalando.org/created1': '{"started":"2021-08-18T20:15:46.545515","stopped":"2021-08-18T20:15:46.548260","purpose":"create","retries":1,"success":true,"failure":false}', 'kopf.zalando.org/created2': '{"started":"2021-08-18T20:15:46.545526","purpose":"create","retries":0,"success":false,"failure":false}'}}, 'status': {'kopf': {'progress': {'created1': {'started': '2021-08-18T20:15:46.545515', 'stopped': '2021-08-18T20:15:46.548260', 'delayed': None, 'purpose': 'create', 'retries': 1, 'success': True, 'failure': False, 'message': None, 'subrefs': None}, 'created2': {'started': '2021-08-18T20:15:46.545526', 'stopped': None, 'delayed': None, 'purpose': 'create', 'retries': 0, 'success': False, 'failure': False, 'message': None, 'subrefs': None}}}}}
[2021-08-18 20:15:46,661] kopf.objects         [DEBUG   ] [orbit-system/image-replication-929js] Creation is in progress: {'apiVersion': 'orbit.aws/v1', 'kind': 'ImageReplication', 'metadata': {'annotations': {'kopf.zalando.org/created1': '{"started":"2021-08-18T20:15:46.545515","stopped":"2021-08-18T20:15:46.548260","purpose":"create","retries":1,"success":true,"failure":false}', 'kopf.zalando.org/created2': '{"started":"2021-08-18T20:15:46.545526","purpose":"create","retries":0,"success":false,"failure":false}'}, 'creationTimestamp': '2021-08-18T20:15:46Z', 'generateName': 'image-replication-', 'generation': 2, 'managedFields': [{'apiVersion': 'orbit.aws/v1', 'fieldsType': 'FieldsV1', 'fieldsV1': {'f:metadata': {'f:annotations': {'.': {}, 'f:kopf.zalando.org/created1': {}, 'f:kopf.zalando.org/created2': {}}}, 'f:status': {'.': {}, 'f:kopf': {'.': {}, 'f:progress': {'.': {}, 'f:created1': {'.': {}, 'f:failure': {}, 'f:purpose': {}, 'f:retries': {}, 'f:started': {}, 'f:stopped': {}, 'f:success': {}}, 'f:created2': {'.': {}, 'f:failure': {}, 'f:purpose': {}, 'f:retries': {}, 'f:started': {}, 'f:success': {}}}}}}, 'manager': 'kopf', 'operation': 'Update', 'time': '2021-08-18T20:15:46Z'}, {'apiVersion': 'orbit.aws/v1', 'fieldsType': 'FieldsV1', 'fieldsV1': {'f:metadata': {'f:generateName': {}}, 'f:spec': {'.': {}, 'f:destination': {}, 'f:source': {}}}, 'manager': 'kubectl-create', 'operation': 'Update', 'time': '2021-08-18T20:15:46Z'}], 'name': 'image-replication-929js', 'namespace': 'orbit-system', 'resourceVersion': '5848832', 'uid': 'a7887a0f-b599-4ab6-8760-e1c8d71a228c'}, 'spec': {'destination': '011316101677.dkr.ecr.us-west-2.amazonaws.com/orbit-dev-env/nginx:latest', 'source': 'nginx:latest'}, 'status': {'kopf': {'progress': {'created1': {'failure': False, 'purpose': 'create', 'retries': 1, 'started': '2021-08-18T20:15:46.545515', 'stopped': '2021-08-18T20:15:46.548260', 'success': True}, 'created2': {'failure': False, 'purpose': 'create', 'retries': 0, 'started': '2021-08-18T20:15:46.545526', 'success': False}}}}}
[2021-08-18 20:15:46,661] kopf.objects         [DEBUG   ] [orbit-system/image-replication-929js] Handler 'created2' is invoked.
[2021-08-18 20:15:46,661] kopf.objects         [INFO    ] [orbit-system/image-replication-929js] Created2
[2021-08-18 20:15:46,663] kopf.objects         [INFO    ] [orbit-system/image-replication-929js] Handler 'created2' succeeded.
[2021-08-18 20:15:46,663] kopf.objects         [INFO    ] [orbit-system/image-replication-929js] Creation is processed: 2 succeeded; 0 failed.
[2021-08-18 20:15:46,664] kopf.objects         [DEBUG   ] [orbit-system/image-replication-929js] Patching with: {'metadata': {'annotations': {'kopf.zalando.org/created2': None, 'kopf.zalando.org/created1': None, 'kopf.zalando.org/last-handled-configuration': '{"spec":{"destination":"011316101677.dkr.ecr.us-west-2.amazonaws.com/orbit-dev-env/nginx:latest","source":"nginx:latest"}}\n'}}, 'status': {'kopf': {'progress': {'created2': None, 'created1': None}}}}
[2021-08-18 20:15:46,776] kopf.objects         [DEBUG   ] [orbit-system/image-replication-929js] Something has changed, but we are not interested (the essence is the same).
[2021-08-18 20:15:46,776] kopf.objects         [DEBUG   ] [orbit-system/image-replication-929js] Handling cycle is finished, waiting for new changes.
```

### Additional information

I have eliminated all other dependencies and reproduced this with a clean Docker image.

**Comment by nolar** [comment]
Thanks for reporting this with details.

> This results in every resource firing an on.resume when the operator is restarted. 

This is the main (and the only) purpose of the resume-handlers. Can you please clarify what is the problem specifically?

---

As for the progress storage:

The per-handler progress is a temporary record of Kopf. Kopf wipes it as soon as one single handling cycle is finished — which happens when all the handlers that are supposed to run did run.

In the first case, when there is only one handler, the whole cycle is finished on the first run, so the progress is not even stored, only the final state (for change detection in the future).

In the second case, only an intermediate state between the unfinished 1st iteration and the finalising 2nd iteration is stored, and is then wiped (overwritten with nulls) when the cycle is over.

This is as intended.

If you want to persist the handler-specific data without wiping them, you have 2 ways:

1. Store something directly from the handlers. E.g. `patch.metadata.annotations['mynotes'] = '...'` in the handler.
2. Implement and configure your own annotations progress storage that does not wipe the intermediate state.

The storages are quite simple to implement; see the built-in storages as an example: https://github.com/nolar/kopf/blob/1.33.0/kopf/_cogs/configs/progress.py#L131-L243 (ignore all the things regarding "markers" for your needs).

Keep in mind that in some operators, people choose to store their state not even in annotations, but in their own databases (e.g. Redis), so neither the progress records nor the last-handled-configuration appear on the resources ever, and the resource remains clear. 

It means that storing the progress in the annotations is simply Kopf's trade-off on not depending on external databases for the sake of simplicity — at the cost of polluting the resources with noisy records. If another way to store intermediate states appears in Kubernetes, Kopf will switch to that and abandon annotations. Operators should not rely on seeing Kopf's internal annotations — they should use their own.

**Comment by chamcca** [comment]
ooh. Thank you for the details and the clarification. This is all a misunderstanding on my part. I misread the docs about the resume handler:

> The resuming handlers are guaranteed to execute only once per operator lifetime for each resource object (except if errors are retried).

I read that (multiple times) as the resume handlers would execute once per lifetime of the _resource_. In other words, I was expecting either the create or resume would execute one and only time for each resource. When I found the resume executing for every resource on operator restart, I blamed it on missing progress.
